lib/hierarchical.h

```diff
@@ -185,81 +161,124 @@ class HierarchicalRadiosity {
+        auto vrad = VertexRadiosityHandleProperty::createIfNotExists(
+            mesh_, "vertex_radiosity");
+        FaceRadiosityHandle frad;
+        auto exists = mesh_.get_property_handle(frad, "face_radiosity");
+        assert(exists);
+        UNUSED(exists);
+
+        for (const auto v : mesh_.vertices()) {
+            Color& avg = vrad[v];
+            avg = Color(0, 0, 0, 1);
+
+            size_t n = 0;
+            for (const auto f : mesh_.vf_range(v)) {
+                avg += mesh_.property(frad, f);
+                n++;
+            }
+            avg /= n;
+        }
     }
 
-    bool is_root(const Quadnode& p) const {
-        return p.tri_id < tree_->num_triangles();
+    void triangulate_t_vertices() {
+        auto rad = FaceRadiosityHandleProperty::createIfNotExists(
+            mesh_, "face_radiosity");
+
+        size_t n_faces = mesh_.n_faces();
+        for (size_t i = 0; i < n_faces; ++i) {
+            auto face = RadiosityMesh::FaceHandle(i);
+            if (has_t_vertex(mesh_, face)) {
+                size_t n = mesh_.n_faces();
+                triangulate_t_vertices(face);
+
+                // copy radiosity
+                const auto& face_rad = rad[face];
+                for (; n < mesh_.n_faces(); ++n) {
+                    auto new_face = RadiosityMesh::FaceHandle(n);
+                    rad[new_face] = face_rad;
+                }
+            }
+        }
     }
 
-    const Triangle& get_triangle(const Quadnode& p) const {
-        if (is_root(p)) {
-            return (*tree_)[p.root_tri_id];
+    void triangulate_t_vertices(RadiosityMesh::FaceHandle face) {
+        size_t n = mesh_.n_faces();
+        auto fully_triangulated = triangulate_t_vertex(mesh_, face);
+        if (n < mesh_.n_faces() && !fully_triangulated) {
+            triangulate_t_vertices(face);
+            triangulate_t_vertices(RadiosityMesh::FaceHandle(n));
         }
-        return subdivided_tris_[p.tri_id - tree_->num_triangles()];
     }
 
+private:
     float estimate_form_factor(const Quadnode& p, const Quadnode& q) const {
-        const Triangle& tri_p = get_triangle(p);
-        const Triangle& tri_q = get_triangle(q);
-
-        const Vec p_midpoint = tri_p.midpoint();
-        const Vec q_midpoint = tri_q.midpoint();
+        const auto p_midpoint = triangle_midpoint(mesh_, p.vs);
+        const auto p_normal = triangle_normal(mesh_, p.vs);
+        const auto q_midpoint = triangle_midpoint(mesh_, q.vs);
 
         const float cos_theta =
-            tri_p.normal * (q_midpoint - p_midpoint).Normalize();
+            p_normal | (q_midpoint - p_midpoint).normalize();
+        assert(!std::isnan(cos_theta));
         if (cos_theta < 0) {
             return 0;
         }
 
-        const float omega_q = solid_angle(p_midpoint, tri_q);
+        const auto& q_a = mesh_.point(q.vs[0]);
+        const auto& q_b = mesh_.point(q.vs[1]);
+        const auto& q_c = mesh_.point(q.vs[2]);
```

Comment:
Face. Das steckt in der Definition in mesh.h. Das ist ein `OpenMesh::FPropHandleT`. `F` steht für Face.